hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py

```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
+        logging.info(f"Loading slide {slide_id} ...")
+        reader = WSIReader(backend="cucim")
+        loader = LoadPandaROId(reader, level=level, margin=margin)
+        try:
             sample = loader(sample)  # load 'image' and 'mask' from disk
-
-            logging.info(f"Tiling slide {slide_id} ...")
-            image_tiles, mask_tiles, tile_locations, occupancies, _ = \
-                generate_tiles(sample, tile_size, occupancy_threshold)
-            n_tiles = image_tiles.shape[0]
-
-            for i in tqdm(range(n_tiles), f"Tiles ({slide_id[:6]}…)", unit="img", disable=not tile_progress):
-                try:
-                    tile_metadata = save_tile(sample, image_tiles[i], mask_tiles[i], tile_locations[i],
-                                              slide_dir)
-                    tile_metadata['occupancy'] = occupancies[i]
-                    tile_metadata['image'] = os.path.join(slide_dir.name, tile_metadata['image'])
-                    tile_metadata['mask'] = os.path.join(slide_dir.name, tile_metadata['mask'])
-                    dataset_row = ','.join(str(tile_metadata[column]) for column in CSV_COLUMNS)
-                    dataset_csv_file.write(dataset_row + '\n')
-                except Exception as e:
-                    tiles_failure += 1
-                    descriptor = get_tile_descriptor(tile_locations[i]) + '\n'
-                    failed_tiles_file.write(descriptor)
-                    traceback.print_exc()
-                    warnings.warn(f"An error occurred while saving tile "
-                                  f"{get_tile_id(slide_id, tile_locations[i])}: {e}")
-
+        except (ValueError, RuntimeError) as e:
+            logging.error(f'Error loading slide {slide_id}:\n{e}')
             dataset_csv_file.close()
-            failed_tiles_file.close()
-            if tiles_failure > 0:
-                # TODO what we want to do with slides that have some failed tiles?
-                logging.warning(f"{slide_id} is incomplete. {tiles_failure} tiles failed.")
-        except Exception as e:
-            traceback.print_exc()
-            warnings.warn(f"An error occurred while processing slide {slide_id}: {e}")
+            shutil.rmtree(slide_dir)
+            return
+        logging.info(f"Tiling slide {slide_id} ...")
+        image_tiles, mask_tiles, tile_boxes, occupancies, num_discarded = \
+            generate_tiles(sample, tile_size, occupancy_threshold)
+        n_tiles = image_tiles.shape[0]
+
+        for i in tqdm(range(n_tiles), f"Tiles ({slide_id[:6]}…)", unit="img", disable=not tile_progress):
+            tile_metadata = save_tile(
+                sample,
+                image_tiles[i],
+                mask_tiles[i],
+                tile_boxes[i],
+                slide_dir,
+            )
+            tile_metadata['occupancy'] = occupancies[i]
+            tile_metadata['image'] = os.path.join(slide_dir.name, tile_metadata['image'])
```

Comment:
[optional] Just noticed this could be simplified to `slide_dir / tile_metadata['image']` (same below), if you felt like it.

Reply:
Not really because `slide_dir` is an absolute path. But I've made some edits to avoid using `os`. Thanks!

Reply:
Ah yes, this is to produce the relative paths that go in the CSV